File: shared-java/src/main/kotlin/com/kyhsgeekcode/minecraftenv/ProtocolVersion.kt
package com.kyhsgeekcode.minecraftenv

/**
 * Bumped whenever InitialEnvironmentMessage/ObservationSpaceMessage/ActionSpaceMessageV2 change in
 * a way that breaks wire compatibility with older Java/Python builds. Exchanged once per session
 * via [MessageIO.writeHandshakeAck]/`HandshakeAck.protocol_version` so an incompatible pairing
 * fails fast on the Python side instead of hanging or misinterpreting bytes deep into a session
 * (docs/26_2_MigrationPlan.md item (f)). Must match Python's
 * `craftground.protocol_version.CRAFTGROUND_PROTOCOL_VERSION`.
 */
const val CRAFTGROUND_PROTOCOL_VERSION = 1

File: src/craftground/protocol_version.py
"""CraftGround IPC protocol handshake (docs/26_2_MigrationPlan.md item (f)).

Bumped whenever InitialEnvironmentMessage/ObservationSpaceMessage/ActionSpaceMessageV2 change in a
way that breaks wire compatibility with older Java/Python builds. Must match shared-java's
com.kyhsgeekcode.minecraftenv.CRAFTGROUND_PROTOCOL_VERSION.
"""

from .proto.initial_environment_pb2 import HandshakeAck

CRAFTGROUND_PROTOCOL_VERSION = 1


class ProtocolVersionMismatchError(RuntimeError):
    """Raised when the Java side's HandshakeAck reports an incompatible protocol_version."""


def validate_handshake_ack(ack: HandshakeAck) -> None:
    if ack.protocol_version != CRAFTGROUND_PROTOCOL_VERSION:
        raise ProtocolVersionMismatchError(
            "CraftGround protocol mismatch: this Python package speaks protocol "
            f"{CRAFTGROUND_PROTOCOL_VERSION}, but the connected Java side (Minecraft "
            f"{ack.minecraft_version}, render backend '{ack.render_backend}') speaks protocol "
            f"{ack.protocol_version}. Reinstall matching craftground and "
            "craftground-runtime-mc121/mc262 versions."
        )
